Mathlib/RingTheory/Bialgebra/TensorProduct.lean

```diff
@@ -0,0 +1,166 @@
+would be welcome. For instance, the tensor product of bialgebras in the
+[FLT repo](https://github.com/ImperialCollegeLondon/FLT/blob/eef74b4538c8852363936dfaad23e6ffba72eca5/FLT/mathlibExperiments/Coalgebra/TensorProduct.lean)
+is already universe polymorphic since it does not go via category theory.
+
+-/
+
+universe v u
+
+open scoped TensorProduct
+
+namespace Bialgebra.TensorProduct
+variable (R A B : Type u) [CommRing R] [Ring A] [Ring B]
+    [Bialgebra R A] [Bialgebra R B]
+
+lemma counit_eq_algHom_toLinearMap :
+    Coalgebra.counit (R := R) (A := A ⊗[R] B) =
+      ((Algebra.TensorProduct.lmul' R).comp (Algebra.TensorProduct.map
+      (Bialgebra.counitAlgHom R A) (Bialgebra.counitAlgHom R B))).toLinearMap := by
+  rfl
+
+lemma comul_eq_algHom_toLinearMap :
+    Coalgebra.comul (R := R) (A := A ⊗[R] B) =
+      ((Algebra.TensorProduct.tensorTensorTensorComm R A A B B).toAlgHom.comp
+      (Algebra.TensorProduct.map (Bialgebra.comulAlgHom R A)
+      (Bialgebra.comulAlgHom R B))).toLinearMap := by
+  rfl
+
+end Bialgebra.TensorProduct
+
+noncomputable instance TensorProduct.instBialgebra
+    {R A B : Type u} [CommRing R] [Ring A] [Ring B]
+    [Bialgebra R A] [Bialgebra R B] : Bialgebra R (A ⊗[R] B) := by
+  have hcounit := congr(DFunLike.coe $(Bialgebra.TensorProduct.counit_eq_algHom_toLinearMap R A B))
+  have hcomul := congr(DFunLike.coe $(Bialgebra.TensorProduct.comul_eq_algHom_toLinearMap R A B))
+  refine Bialgebra.mk' R (A ⊗[R] B) ?_ (fun {x y} => ?_) ?_ (fun {x y} => ?_) <;>
+  simp_all only [AlgHom.toLinearMap_apply] <;>
+  simp only [_root_.map_one, _root_.map_mul]
+
+namespace Bialgebra.TensorProduct
+
+variable {R A B C D : Type u} [CommRing R] [Ring A] [Ring B] [Ring C] [Ring D]
+    [Bialgebra R A] [Bialgebra R B] [Bialgebra R C] [Bialgebra R D]
+
+/-- The tensor product of two bialgebra morphisms as a bialgebra morphism. -/
+noncomputable def map (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    A ⊗[R] C →ₐc[R] B ⊗[R] D :=
+  { Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D),
+    Algebra.TensorProduct.map (f : A →ₐ[R] B) (g : C →ₐ[R] D) with }
+
+@[simp]
+theorem map_tmul (f : A →ₐc[R] B) (g : C →ₐc[R] D) (x : A) (y : C) :
+    map f g (x ⊗ₜ y) = f x ⊗ₜ g y :=
+  rfl
+
+@[simp]
+theorem map_toCoalgHom (f : A →ₐc[R] B) (g : C →ₐc[R] D) :
+    map f g = Coalgebra.TensorProduct.map (f : A →ₗc[R] B) (g : C →ₗc[R] D) := rfl
+
+variable (R A B C)
+
+/-- The associator for tensor products of R-bialgebras, as a bialgebra equivalence. -/
+protected noncomputable def assoc :
+    (A ⊗[R] B) ⊗[R] C ≃ₐc[R] A ⊗[R] (B ⊗[R] C) :=
+  { Coalgebra.TensorProduct.assoc R A B C, Algebra.TensorProduct.assoc R A B C with }
+
+variable {R A B C}
+
+@[simp]
+theorem assoc_tmul (x : A) (y : B) (z : C) :
+    Bialgebra.TensorProduct.assoc R A B C ((x ⊗ₜ y) ⊗ₜ z) = x ⊗ₜ (y ⊗ₜ z) :=
+  rfl
+
+@[simp]
+theorem assoc_symm_tmul (x : A) (y : B) (z : C) :
+    (Coalgebra.TensorProduct.assoc R A B C).symm (x ⊗ₜ (y ⊗ₜ z)) = (x ⊗ₜ y) ⊗ₜ z :=
+  rfl
```

Comment:
This is a theorem about coalgebras in the bialgebra namespace. I'm assuming this wasn't intentional?

Reply:
Indeed it's a mistake, thanks!